From: Dermot Bradley <dermot_bradley@yahoo.com>
Date: Wed, 8 Jun 2022 18:28:13 +0100
Subject: [PATCH] remove use of udevadm

These changes in this patch are copied from upstream's master repo
based on commit 1ea4e5ef09c06552402bf676ce262a661372f08d.

Replace the use of eudev's udevadm by grub-install/grub-probe with
the use of the /sys/dev/block/ directory hierarchy instead. This
patch enables grub-install (via grub-probe) to install onto a
partitioned loop device inside a chroot.

Currently when grub-probe runs to detect the device containing the
boot filesystem it uses udevadm, if available, to do so and if not
available it falls back to its own logic which cannot handle a
partitioned loop device inside a chroot.

Without this patch the only current workaround for a Alpine system
using mdev in such a scenario is to temporarily install eudev (for
the udevadm binary) prior to running grub-install and then to
uninstall eudev afterwards.

Based on upstream's timescale for releases it is likely that the next
Grub release will be some time away in the future and so it would be
useful to backport this change to Alpine.

---

diff -Naur a/grub-core/osdep/linux/hostdisk.c b/grub-core/osdep/linux/hostdisk.c
--- a/grub-core/osdep/linux/hostdisk.c
+++ b/grub-core/osdep/linux/hostdisk.c
@@ -31,6 +31,7 @@
 #include <grub/misc.h>
 #include <grub/i18n.h>
 #include <grub/list.h>
+#include <grub/osdep/major.h>
 
 #include <stdio.h>
 #include <stdlib.h>
@@ -98,54 +99,13 @@
 static char *
 sysfs_partition_path (const char *dev, const char *entry)
 {
-  const char *argv[7];
-  int fd;
-  pid_t pid;
-  FILE *udevadm;
-  char *buf = NULL;
-  size_t len = 0;
-  char *path = NULL;
-
-  argv[0] = "udevadm";
-  argv[1] = "info";
-  argv[2] = "--query";
-  argv[3] = "path";
-  argv[4] = "--name";
-  argv[5] = dev;
-  argv[6] = NULL;
-
-  pid = grub_util_exec_pipe (argv, &fd);
-
-  if (!pid)
-    return NULL;
-
-  /* Parent.  Read udevadm's output.  */
-  udevadm = fdopen (fd, "r");
-  if (!udevadm)
-    {
-      grub_util_warn (_("Unable to open stream from %s: %s"),
-		      "udevadm", strerror (errno));
-      close (fd);
-      goto out;
-    }
-
-  if (getline (&buf, &len, udevadm) > 0)
-    {
-      char *newline;
-
-      newline = strchr (buf, '\n');
-      if (newline)
-	*newline = '\0';
-      path = xasprintf ("/sys%s/%s", buf, entry);
-    }
-
-out:
-  if (udevadm)
-    fclose (udevadm);
-  waitpid (pid, NULL, 0);
-  free (buf);
+  struct stat st;
 
-  return path;
+  if (stat (dev, &st) == 0 && S_ISBLK (st.st_mode))
+    return xasprintf ("/sys/dev/block/%u:%u/%s",
+		      major (st.st_rdev), minor (st.st_rdev), entry);
+
+  return NULL;
 }
 
 static int
diff -Naur a/include/grub/osdep/major.h b/include/grub/osdep/major.h
--- a/include/grub/osdep/major.h
+++ b/include/grub/osdep/major.h
@@ -0,0 +1,33 @@
+/*
+ *  GRUB  --  GRand Unified Bootloader
+ *  Copyright (C) 2021  Free Software Foundation, Inc.
+ *
+ *  GRUB is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  GRUB is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  Fix for glibc 2.25 which is deprecating the namespace pollution of
+ *  sys/types.h injecting major(), minor(), and makedev() into the
+ *  compilation environment.
+ */
+
+#ifndef GRUB_OSDEP_MAJOR_H
+#define GRUB_OSDEP_MAJOR_H	1
+
+#include <sys/types.h>
+
+#ifdef MAJOR_IN_MKDEV
+# include <sys/mkdev.h>
+#elif defined (MAJOR_IN_SYSMACROS)
+# include <sys/sysmacros.h>
+#endif
+#endif /* GRUB_OSDEP_MAJOR_H */
